src/crypto/tls/delegated_credentials_test.go

```diff
@@ -0,0 +1,413 @@
+	if err != nil {
+		panic(err)
+	}
+	dcTest = &dcAndPrivateKey{dc, sk}
+}
+
+// TODO: generalize for all schemes
+func publicKeysEqual(publicKey, publicKey2 crypto.PublicKey, algo SignatureScheme) error {
+	curve := getCurve(algo)
+	pk := publicKey.(*ecdsa.PublicKey)
+	pk2 := publicKey2.(*ecdsa.PublicKey)
+
+	serPubKey := elliptic.Marshal(curve, pk.X, pk.Y)
+	serPubKey2 := elliptic.Marshal(curve, pk2.X, pk2.Y)
+	if !bytes.Equal(serPubKey2, serPubKey) {
+		return errors.New("Public Keys mismatch")
+	}
+
+	return nil
+}
+
+func equal(dc, dc2 *DelegatedCredential) error {
+	if dc2.Cred.ValidTime != dc.Cred.ValidTime {
+		return fmt.Errorf("ValidTime mismatch: got %d; want %d", dc2.Cred.ValidTime, dc.Cred.ValidTime)
+	}
+
+	if dc2.Cred.expCertVerfAlgo != dc.Cred.expCertVerfAlgo {
+		return fmt.Errorf("scheme mismatch: got %04x; want %04x", dc2.Cred.expCertVerfAlgo, dc.Cred.expCertVerfAlgo)
+	}
+
+	return publicKeysEqual(dc.Cred.PublicKey, dc2.Cred.PublicKey, dc.Cred.expCertVerfAlgo)
+}
+
+// Test delegation and validation of credentials.
+func TestDelegateCredentialsValidate(t *testing.T) {
+	initialize()
+	cert := dcTestCerts["dc"]
+	validTime := dcNow.Sub(cert.Leaf.NotBefore) + dcMaxTTL
+
+	delegatedCred, _, err := NewDelegatedCredential(cert, ECDSAWithP256AndSHA256, validTime, "server")
+	if err != nil {
+		t.Fatal(err)
+	} else if delegatedCred == nil {
+		t.Fatal("unable to generate a Delegated Credential")
+	}
+
+	// Valid Delegated Credential
+	if !delegatedCred.Validate(cert.Leaf, "server", dcNow) {
+		t.Error("generated valid Delegated Credential is rendered invalid")
+	}
+
+	// Expired Delegated Credential
+	expired := dcNow.Add(dcMaxTTL).Add(time.Nanosecond)
+	if delegatedCred.Validate(cert.Leaf, "server", expired) {
+		t.Error("expired delegated credential is valid; want invalid")
+	}
+
+	// Test validation of Delegated Credential which TTL is too long
+	invalidDelegatedCred, _, err := NewDelegatedCredential(cert, ECDSAWithP256AndSHA256, validTime+time.Second, "server")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if invalidDelegatedCred.Validate(cert.Leaf, "server", dcNow) {
+		t.Error("Delegated Credential validation with long TTL succeeded; want failure")
+	}
+
+	shortValidTime := dcNow.Sub(cert.Leaf.NotBefore) + time.Second
+
+	// Test validation of Delegated Credential which TTL is short
+	delegatedCred, _, err = NewDelegatedCredential(cert, ECDSAWithP256AndSHA256, shortValidTime, "server")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !delegatedCred.Validate(cert.Leaf, "server", dcNow) {
+		t.Error("valid Delegated Credential is invalid; want valid")
+	}
+
+	//cert.Leaf.SignatureAlgorithm = x509.ECDSAWithSHA256
+	//delegatedCred.Algorithm = ECDSAWithP521AndSHA512
+
+	// Test signature algorithm binding
+	//if delegatedCred.Validate(cert.Leaf, "server", dcNow) {
+	//	t.Error("Delegated Credential with wrong scheme is valid; want invalid")
+	//}
```

Comment:
```suggestion
```

(dead code)

Reply:
Answered in another comment.

Reply:
Which other comment? Let's not commit dead code.